src/canister_tests/src/certificate_validation.rs

```diff
@@ -0,0 +1,145 @@
+// adapted from https://github.com/dfinity/icx-proxy/blob/0cd1a22f717b56ac550a3554a25a845878bfb4e8/src/main.rs#L611
+// TODO: certificate validation should be it's own library
+
+use crate::certificate_validation::ValidationError::{
+    AssetHashMismatch, AssetPathLookupFailed, MalformedCertificate, TreeHashCertifiedDataMismatch,
+    WitnessLookupFailed,
+};
+use candid::types::ic_types::hash_tree::LookupResult;
+use flate2::read::GzDecoder;
+use ic_agent::agent::http_transport::ReqwestHttpReplicaV2Transport;
+use ic_agent::hash_tree::HashTree;
+use ic_agent::{lookup_value, Agent, AgentError};
+use ic_crypto_utils_threshold_sig_der::public_key_to_der;
+use ic_state_machine_tests::{CanisterId, ThresholdSigPublicKey};
+use regex::Regex;
+use serde::Deserialize;
+use sha2::{Digest, Sha256};
+use std::io::Read;
+
+// The limit of a buffer we should decompress ~10mb.
+const MAX_CHUNK_SIZE_TO_DECOMPRESS: usize = 1024;
+const MAX_CHUNKS_TO_DECOMPRESS: u64 = 10_240;
+
+const CERTIFICATE_REGEX: &'static str = "^certificate=:([^:]*):, tree=:([^:]*):$";
+
+#[derive(Debug)]
+pub enum ValidationError {
+    MalformedCertificate { message: String },
+    CertificateValidationFailed { inner: AgentError },
+    WitnessLookupFailed { inner: AgentError },
+    TreeHashCertifiedDataMismatch,
+    AssetPathLookupFailed,
+    AssetHashMismatch,
+}
+
+pub fn validate_certification(
+    ic_certificate: &str,
+    canister_id: CanisterId,
+    uri_path: &str,
+    body: &[u8],
+    encoding: Option<&str>,
+    root_key: ThresholdSigPublicKey,
+) -> Result<(), ValidationError> {
+    let cert = parse_header_value(ic_certificate, 1)?;
+    let tree: HashTree = parse_header_value(ic_certificate, 2)?;
+    let agent = initialize_agent(&root_key);
```

Comment:
what is the role of the "agent"? is there any chance we can steal some code from ic-agent instead of tying us to a new dependency?

Reply:
I tried, but it's worse than having a dependency on `agent-rs`. The reason is that `agent-rs` includes a _modified_ copy of https://github.com/miracl/core. And the reason for that, in turn, is that the library can only be configured by copying and modifying it. 🙃

Reply:
😭